### PR TITLE
dynamic proof aggregation

### PR DESCRIPTION
**Pull request by zhenfeizhang**
### Description

Dynamic proof aggregation

### Issue Link

[_link issue here_]

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

Dynamic proof aggregation. See https://github.com/scroll-tech/zkevm-circuits/blob/dynamic-proof-aggregation/aggregator/README.md for the spec

### Rationale

[_design decisions and extended information_]

### How Has This Been Tested?
Quick test
```
 cargo test --release --features=disable_proof_aggregation
```

Full test log:
Super: /home/ubuntu/zhenfei/0227/zkevm-circuits/aggregator/e2e.log

### CODE REVIEW

**Comment by zhenfeizhang** [comment]
@lispc 
CI requires an update to include the tests in this crate

```
 cargo test --release --features=disable_proof_aggregation
```

default test without this feature flag will take too long

File: aggregator/src/rlc.rs [hunk 1]
```diff
@@ -0,0 +1,4 @@
+mod config;
```
Comment (icemelon) [inline aggregator/src/rlc.rs:1]: I suggest move `rlc` module inside the `aggregation` modules as it is a gadget used by aggregation circuit only.

Comment (zhenfeizhang) [reply aggregator/src/rlc.rs:1]: will fix after steven's integration test passes

File: aggregator/README.md [hunk 2]
```diff
@@ -56,4 +56,197 @@ In addition, it attests that, for chunks indexed from `0` to `k-1`,
...
+
+Currently `n` is hard coded to `10`.
+# Structs
+
+## Chunk
+
+A __chunk__ is a list of continuous blocks. It consists of 4 hashes:
+- state root before this chunk
+- state root after this chunk
+- the withdraw root of this chunk
```
Comment (lispc) [inline aggregator/README.md:78]: the withdraw root of this chunk => the withdraw root after this chunk?

File: aggregator/src/aggregation/circuit.rs [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+
+        let timer = start_timer!(|| ("assign hash cells").to_string());
+        let (hash_digest_cells, num_valid_snarks) = assign_batch_hashes(
+            &config,
+            &mut layouter,
+            challenges,
+            &preimages,
+            self.batch_hash.number_of_valid_chunks,
+        )
+        .map_err(|_e| Error::ConstraintSystemFailure)?;
```
Comment (lispc) [inline aggregator/src/aggregation/circuit.rs]: TODO(zhuo): check this error handling here

File: aggregator/src/aggregation/config.rs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            };
+
+            KeccakCircuitConfig::new(meta, keccak_circuit_config_args)
+        };
+
+        // The current code base is hardcoded for KeccakCircuit configured
+        // with 300 rows and 87 columns per hash call.
+        let columns = keccak_circuit_config.cell_manager.columns();
+
+        assert_eq!(
```
Comment (lispc) [inline aggregator/src/aggregation/config.rs:90]: TODO(zhuo): check the 87 columns

File: aggregator/src/core.rs [hunk 5]
```diff
@@ -62,3 +91,601 @@ pub(crate) fn extract_accumulators_and_proof(
...
+                        // 10-th column is Keccak input in Keccak circuit
+                        hash_input_cells.push(row[10].clone());
+                        cur_preimage_index = preimage_indices_iter.next();
+                    }
+                    if cur_digest_index.is_some() && *cur_digest_index.unwrap() == offset {
+                        // last column is Keccak output in Keccak circuit
+                        hash_output_cells.push(row.last().unwrap().clone()); // sage unwrap
+                        cur_digest_index = digest_indices_iter.next();
+                    }
+                    if offset % 300 == 0 && offset / 300 < 30 {
```
Comment (lispc) [inline aggregator/src/core.rs:232]: replace with 25 * KECCAK_ROWS?

File: aggregator/src/core.rs [hunk 6]
```diff
@@ -62,3 +91,601 @@ pub(crate) fn extract_accumulators_and_proof(
...
+                let flag3 = assigned_value_to_cell(
+                    rlc_config,
+                    &mut region,
+                    &data_hash_flag_cells[2],
+                    &mut offset,
+                )?;
+
+                println!("flag1: {:?}", flag1.value());
+                println!("flag2: {:?}", flag2.value());
+                println!("flag3: {:?}", flag3.value());
```
Comment (lispc) [inline aggregator/src/core.rs:546]: comment or log trace

**Comment by lispc** [comment]
what is the column number of the final layer circuit? 87(keccak) + the flex_gate columns?    Did you notice the gas cost of solidity verification?

File: aggregator/src/util.rs [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+
+use crate::{
+    constants::{DIGEST_LEN, MAX_AGG_SNARKS, ROUND_LEN},
+    rlc::RlcConfig,
+    DEFAULT_KECCAK_ROWS, NUM_ROUNDS,
+};
+
+use std::env::var;
+
+pub(crate) fn capacity(num_rows: usize) -> Option<usize> {
```
Comment (lispc) [inline aggregator/src/util.rs:21]: rename like keccak_round_capacity?

**Comment by zhenfeizhang** [comment]
replaced with #670 Static proof aggregation

**Comment by zhenfeizhang** [comment]
closed as #670 is merged

**Pull request closed without merging**